### PR TITLE
Add subgraph fuse support and anakin engine

### PR DESCRIPTION
**Pull request by NHZlX**

### CODE REVIEW

File: paddle/fluid/inference/analysis/ir_pass_manager.cc [hunk 1]
```diff
@@ -62,6 +65,11 @@ void IRPassManager::CreatePasses(Argument *argument,
                     argument->mkldnn_enabled_op_types()));
     }
 
+    if (pass_name == "anakin_subgraph_pass") {
+      pass->Set("program",
+                new framework::ProgramDesc *(&argument->main_program()));
```
Comment (Superjomn) [inline paddle/fluid/inference/analysis/ir_pass_manager.cc:70]: the main_program is the original program,
it is also stored in the graph passed to Graph.

File: paddle/fluid/inference/analysis/ir_passes/anakin_subgraph_pass.cc [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+  auto *op_desc = node->Op();
+  auto &subgraph = *Agent(node).subgraph();
+  PADDLE_ENFORCE(!subgraph.empty());
+
+  framework::ProgramDesc *program_desc =
+      Get<framework::ProgramDesc *>("program");
+  // Add new block for TensorRTEngineOP
+  const framework::BlockDesc &main_block =
+      program_desc->Block(framework::kRootBlockIndex);
+  // const framework::BlockDesc& main_block = program_desc->Block(0);
```
Comment (Superjomn) [inline paddle/fluid/inference/analysis/ir_passes/anakin_subgraph_pass.cc:96]: remove this line

File: paddle/fluid/operators/anakin/anakin_engine_op.h [hunk 3]
```diff
@@ -0,0 +1,152 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (Superjomn) [inline paddle/fluid/operators/anakin/anakin_engine_op.h]: LIcense format is wrong

File: paddle/fluid/inference/api/paddle_pass_builder.cc [hunk 4]
```diff
@@ -70,23 +70,29 @@ void GpuPassStrategy::EnableMKLDNN() {
...
-        "conv_elementwise_add_act_fuse_pass",   //
-        "conv_elementwise_add2_act_fuse_pass",  //
-        "conv_elementwise_add_fuse_pass",       //
-#endif
+      "infer_clean_graph_pass",  //
+      "anakin_subgraph_pass",
+      //     "identity_scale_op_clean_pass",              //
+      /*
+             "conv_affine_channel_fuse_pass",             //
+             "conv_eltwiseadd_affine_channel_fuse_pass",  //
```
Comment (Superjomn) [inline paddle/fluid/inference/api/paddle_pass_builder.cc:78]: Need to add an API to switch these latter

File: paddle/fluid/operators/anakin/anakin_engine_op.h [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    block_desc.ParseFromString(Attr<std::string>("subgraph"));
+
+    std::vector<std::string> output_maps =
+        Attr<std::vector<std::string>>("output_name_mapping");
+
+    inference::Singleton<inference::anakin::AnakinOpConverter>::Global()
+        .ConvertBlock(block_desc, param_names_, scope, engine);
+    engine->Freeze();
+    engine->Optimize();
+    engine->Save("anakin.saved");
```
Comment (Superjomn) [inline paddle/fluid/operators/anakin/anakin_engine_op.h]: Need to remove this?

File: paddle/fluid/operators/anakin/anakin_engine_op.h [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+    for (const auto &x : Inputs("Xs")) {
+      if (param_names_.count(x)) continue;
+      auto &t =
+          inference::analysis::GetFromScope<framework::LoDTensor>(scope, x);
+      auto t_shape = framework::vectorize2int(t.dims());
+      std::cout << "!!!!!!!!" << t_shape.size() << std::endl;
```
Comment (Superjomn) [inline paddle/fluid/operators/anakin/anakin_engine_op.h]: remove these, or  use VLOG

File: paddle/fluid/inference/anakin/convert/fc.cc [hunk 7]
```diff
@@ -54,18 +67,54 @@ void FcOpConverter::operator()(const framework::proto::OpDesc &op,
   framework::LoDTensor weight_tensor;
   weight_tensor.Resize(y_t->dims());
   TensorCopySync((*y_t), platform::CPUPlace(), &weight_tensor);
+  auto *weight_data = weight_tensor.data<float>();
+  PADDLE_ENFORCE(w_m * w_k == weight_tensor.numel());
```
Comment (fc500110) [inline paddle/fluid/inference/anakin/convert/fc.cc:71]: weight_tensor.numel()返回类型是int64_t

**Review by fc500110 — APPROVED** [review]
LGTM

**Pull request merged**